include/rabit/internal/utils.h

```diff
@@ -40,6 +40,40 @@
 #define fopen64 std::fopen
 #endif  // __APPLE__
 
+// keeps rabit api caller signature
+#ifndef RABIT_API_CALLER_SIGNATURE
+#define RABIT_API_CALLER_SIGNATURE
+
+#ifdef __has_builtin
+
+#if __has_builtin(__builtin_FILE)
+#define _FILE  __builtin_FILE()
+#else
+#define _FILE  "N/A"
+#endif  // __has_builtin(__builtin_FILE)
+
+#if __has_builtin(__builtin_LINE)
+#define _LINE  __builtin_LINE()
```

Comment:
does that mean if we wrap init boostrap in a helper function and call the function with a different higher level caller, it will fail?

Reply:
if you were referring someone wrap boostrap allreduce in helper function and call it more than once. It might **fail** but not unexpected behavior.  e.g Dmatrix::Load of training and testing datasets which has exactly same signature. detail how signature were calculated below.

![image](https://user-images.githubusercontent.com/1184534/63609843-d881f680-c58c-11e9-9c76-eee5149d5f92.png)

when user see failure to run tests, it can follow following way to resolve (pass down unique key from upper function )
![image](https://user-images.githubusercontent.com/1184534/63609944-11ba6680-c58d-11e9-9add-9c89eb504af9.png)

Reply:
update we remove this in upcoming pr
https://github.com/dmlc/rabit/pull/102/commits/69789ec4510d019d25127c62973acef10d10c6fc